ide/api.lsp/src/org/netbeans/spi/lsp/CompletionCollector.java

```diff
@@ -0,0 +1,250 @@
+ * in MimeLookup.
+ *
+ * @author Dusan Balek
+ */
+@MimeLocation(subfolderName = "CompletionCollectors")
+public interface CompletionCollector {
+
+    /**
+     * Computes and collects completions for a document at a given offset.
+     * This method is called outside of AWT to collect completions and
+     * send them via the Language Server Protocol to client for display.
+     *
+     * @param doc a text document
+     * @param offset an offset inside the text document
+     * @param context an optional completion context
+     * @param consumer an operation accepting collected completions
+     *
+     * @return true if the list of collected completion is complete
+     */
+    public boolean collectCompletions(@NonNull Document doc, int offset, @NullAllowed Completion.Context context, @NonNull Consumer<Completion> consumer);
+
+    /**
+     * Creates a builder for {@link Completion} instances
+     *
+     * @param label the label of the completion
+     * @return newly created builder
+     */
+    public static Builder newBuilder(@NonNull String label) {
+        return new Builder(label);
+    }
+
+    public static final class Builder {
+
+        private String label;
+        private Completion.Kind kind;
+        private List<Completion.Tag> tags;
+        private CompletableFuture<String> detail;
+        private CompletableFuture<String> documentation;
+        private boolean preselect;
+        private String sortText;
+        private String filterText;
+        private String insertText;
+        private Completion.TextFormat insertTextFormat;
+        private TextEdit textEdit;
+        private CompletableFuture<List<TextEdit>> additionalTextEdits;
+        private List<Character> commitCharacters;
+        private Command command;
+
+        private Builder(@NonNull String label) {
+            this.label = label;
+        }
+
+        /**
+         * The label of this completion. By default also the text that is inserted
+         * when selecting this completion.
+         */
+        @NonNull
+        public Builder label(@NonNull String label) {
+            this.label = label;
+            return this;
+        }
+
+        /**
+         * The kind of this completion.
+         */
+        @NonNull
+        public Builder kind(@NonNull Completion.Kind kind) {
+            this.kind = kind;
+            return this;
+        }
+
+        /**
+         * Adds tag for this completion.
+         */
+        @NonNull
+        public Builder addTag(@NonNull Completion.Tag tag) {
+            if (this.tags == null) {
+                this.tags = new ArrayList<>();
+            }
+            this.tags.add(tag);
+            return this;
+        }
+
+        /**
+         * A human-readable string with additional information
+         * about this completion, like type or symbol information.
+         */
+        @NonNull
+        public Builder detail(@NonNull CompletableFuture<String> detail) {
```

Comment:
I complained about use of `CompletableFuture` in the API and here I am going to complain as well. In my opinion here we don't want `CompletableFuture` at all. Ideally we want to register _some computation_ that gets triggered when `Completion.getDetail()` is called. My favorite signature is:
```
public Builder detail(Executor whereToRunTheSupplier, Supplier<String> computeTheValueWhenNeeded);
```
the API part would then be
```
CompletionStage<String> getDetail() {
  return CompletableFuture.supplyAsync(computeTheValueWhenNeeded, whereToRunTheSupplier);
}
```
preferrably without allowing one to cast to `CompletableFuture` and also keeping the first instances, once computation is started. What do you think @sdedic?

Reply:
Two versions of the `detail` method provided:
```
public Builder detail(String);
```
for immediate setting the value and
```
public Builder detail(Supplier<String>);
```
for a deferred lazy computation of the value.